sled-agent/config-reconciler/src/ledger.rs

```diff
@@ -182,6 +246,891 @@ struct LedgerTask<T> {
+                        true
+                    }
+                });
+
+                Ok(())
+            }
+            Err(err) => {
+                warn!(
+                    self.log, "Failed to write sled config to ledger";
+                    "generation" => %ledger.data().generation,
+                    InlineErrorChain::new(&err),
+                );
+                Err(LedgerNewConfigError::LedgerCommitFailed(err))
+            }
+        }
+    }
+
+    fn validate_artifact_config(
+        &mut self,
+        new_config: ArtifactConfig,
+    ) -> Result<(), LedgerArtifactConfigError> {
+        // It's okay to hold this borrow for the check below; this is quick and
+        // synchronous.
+        let current_config = self.current_config_tx.borrow();
+        let sled_config = match current_config.deref() {
+            CurrentSledConfig::WaitingForInternalDisks => {
+                // If we have no disks, we don't know yet whether we have a sled
+                // config, so it's not safe to accept any artifact config
+                // changes.
+                return Err(LedgerArtifactConfigError::NoInternalDisks);
+            }
+            CurrentSledConfig::WaitingForInitialConfig => {
+                // If we have no config, any artifact config is fine.
+                return Ok(());
+            }
+            CurrentSledConfig::Ledgered(sled_config) => sled_config,
+        };
+
+        let mut missing = BTreeMap::new();
+        for zone in &sled_config.zones {
+            if let Some(hash) = zone.image_source.artifact_hash() {
+                if !new_config.artifacts.contains(&hash) {
+                    missing.insert(hash, "current zone configuration");
+                }
+            }
+        }
+
+        if missing.is_empty() {
+            Ok(())
+        } else {
+            Err(LedgerArtifactConfigError::InUseArtifactedMissing(missing))
+        }
+    }
+
+    async fn wait_for_m2_disks(&mut self) -> Result<(), LedgerTaskExit> {
+        loop {
+            let internal_disks = self.internal_disks_rx.current_and_update();
+            let config_datasets =
+                internal_disks.all_config_datasets().peekable();
+
+            // Check for a sled config on all our config datasets. (If we have
+            // no config datasets, we'll keep waiting!)
+            match load_sled_config(
+                &config_datasets.collect::<Vec<_>>(),
+                &self.log,
+            )
+            .await
+            {
+                // If we're still waiting, fall through to the `select!` below.
+                CurrentSledConfig::WaitingForInternalDisks => (),
+                // Otherwise, we're done waiting: set our loaded config (which
+                // might be `WaitingForInitialConfig` if we have disks but they
+                // have no ledger contents!).
+                config => {
+                    self.current_config_tx.send_modify(|c| *c = config);
```

Comment:
If `wait_for_m2_disks` is called repeatedly, this will update the value in the watch::Sender again, right? Is that fine?

Reply:
It isn't and can't really be called repeatedly; added a comment and moved an assertion in 741152aab. Does that work?

Reply:
Gonna go ahead and merge, but if this is still ugly I'm happy to change it in a future PR

Reply:
I'm fine with this; makes it more clear to me - and the panic seems reasonable in a non-pub fn